File: src/main/java/org/luxons/sevenwonders/game/data/serializers/ScienceProgressSerializer.java
package org.luxons.sevenwonders.game.data.serializers;

import java.lang.reflect.Type;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import org.luxons.sevenwonders.game.boards.Science;
import org.luxons.sevenwonders.game.boards.ScienceType;
import org.luxons.sevenwonders.game.effects.ScienceProgress;

public class ScienceProgressSerializer implements JsonSerializer<ScienceProgress>, JsonDeserializer<ScienceProgress> {

    @Override
    public JsonElement serialize(ScienceProgress scienceProgress, Type typeOfSrc, JsonSerializationContext context) {
        Science science = scienceProgress.getScience();

        if (science.size() > 1) {
            throw new UnsupportedOperationException("Cannot serialize science containing more than one element");
        }

        for (ScienceType type : ScienceType.values()) {
            int quantity = science.getQuantity(type);
            if (quantity == 1) {
                return context.serialize(type);
            }
        }

        if (science.getJokers() == 1) {
            return new JsonPrimitive("any");
        }

        return JsonNull.INSTANCE;
    }

    @Override
    public ScienceProgress deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws
            JsonParseException {
        String s = json.getAsString();
        ScienceProgress scienceProgress = new ScienceProgress();
        Science science = new Science();
        if ("any".equals(s)) {
            science.addJoker(1);
        } else {
            science.add(context.deserialize(json, ScienceType.class), 1);
        }
        scienceProgress.setScience(science);
        return scienceProgress;
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/game/data/serializers/ScienceProgressSerializer.kt
package org.luxons.sevenwonders.game.data.serializers

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonNull
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import org.luxons.sevenwonders.game.boards.Science
import org.luxons.sevenwonders.game.boards.ScienceType
import org.luxons.sevenwonders.game.effects.ScienceProgress
import java.lang.reflect.Type

internal class ScienceProgressSerializer : JsonSerializer<ScienceProgress>, JsonDeserializer<ScienceProgress> {

    override fun serialize(
        scienceProgress: ScienceProgress,
        typeOfSrc: Type,
        context: JsonSerializationContext
    ): JsonElement {
        val science = scienceProgress.science

        if (science.size() > 1) {
            throw UnsupportedOperationException("Cannot serialize science containing more than one element")
        }

        for (type in ScienceType.values()) {
            val quantity = science.getQuantity(type)
            if (quantity == 1) {
                return context.serialize(type)
            }
        }

        return if (science.jokers == 1) JsonPrimitive("any") else JsonNull.INSTANCE
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): ScienceProgress {
        val s = json.asString
        val science = Science()
        if ("any" == s) {
            science.addJoker(1)
        } else {
            science.add(deserializeScienceType(json, context), 1)
        }
        return ScienceProgress(science)
    }

    private fun deserializeScienceType(json: JsonElement, context: JsonDeserializationContext): ScienceType {
        return context.deserialize<ScienceType>(json, ScienceType::class.java)
                ?: throw IllegalArgumentException("Invalid science level " + json.asString)
    }
}
